File: flame/src/main/java/org/movieos/flame/fragments/ServiceListFragment.java
package org.movieos.flame.fragments;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.movieos.flame.MyApplication;
import org.movieos.flame.R;
import org.movieos.flame.databinding.CellWithImageBinding;
import org.movieos.flame.databinding.RecyclerViewFragmentBinding;
import org.movieos.flame.models.FlameHost;
import org.movieos.flame.utilities.BindingViewHolder;
import org.movieos.flame.utilities.DiscoveryService;
import timber.log.Timber;

import java.util.ArrayList;
import java.util.List;

import javax.jmdns.ServiceEvent;

public class ServiceListFragment extends Fragment {

    private static final String HOST_IDENTIFIER = "hostIdentifier";

    final ServiceListAdapter mAdapter = new ServiceListAdapter();
    String mHostIdentifier;

    @Nullable
    private RecyclerViewFragmentBinding mBinding;

    public static Fragment build(final FlameHost host) {
        Fragment fragment = new ServiceListFragment();
        fragment.setArguments(new Bundle());
        fragment.getArguments().putString(HOST_IDENTIFIER, host.getIdentifer());
        return fragment;
    }

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mHostIdentifier = getArguments().getString(HOST_IDENTIFIER);
        if (mHostIdentifier == null) {
            Timber.e("no identifier in bundle");
        }
    }

    @Nullable
    @Override
    public View onCreateView(final LayoutInflater inflater, @Nullable final ViewGroup container, @Nullable final Bundle savedInstanceState) {
        mBinding = RecyclerViewFragmentBinding.inflate(inflater, container, false);
        if (mBinding == null) {
            return null;
        }

        mBinding.toolbar.setTitle(mHostIdentifier);
        Drawable back = ContextCompat.getDrawable(getContext(), R.drawable.ic_arrow_back_black_24dp);
        back.setTint(0xFFFFFFFF);
        mBinding.toolbar.setNavigationIcon(back);
        mBinding.toolbar.setNavigationOnClickListener(v -> getActivity().onBackPressed());

        mBinding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        mBinding.recyclerView.setAdapter(mAdapter);

        DividerItemDecoration divider = new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL);
        divider.setDrawable(ContextCompat.getDrawable(getContext(), R.drawable.divider));
        mBinding.recyclerView.addItemDecoration(divider);

        return mBinding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mBinding != null) {
            mBinding.unbind();
            mBinding = null;
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        onMessageEvent(null);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(DiscoveryService.HostsChangedNotification event) {
        List<FlameHost> hosts = MyApplication.discoveryService().getHostsMatchingKey(mHostIdentifier);
        if (!hosts.isEmpty()) {
            mAdapter.setServices(hosts.get(0).getServices());
        }
    }

    private static class ServiceListAdapter extends RecyclerView.Adapter<BindingViewHolder<CellWithImageBinding>> {

        private final List<ServiceEvent> mServices = new ArrayList<>();

        ServiceListAdapter() {
            super();
        }

        void setServices(List<ServiceEvent> services) {
            mServices.clear();
            mServices.addAll(services);
            notifyDataSetChanged();
        }

        @Override
        public int getItemCount() {
            return mServices.size();
        }

        @Override
        public BindingViewHolder<CellWithImageBinding> onCreateViewHolder(final ViewGroup parent, final int viewType) {
            return BindingViewHolder.build(parent, R.layout.cell_with_image);
        }

        @Override
        public void onBindViewHolder(final BindingViewHolder<CellWithImageBinding> holder, final int position) {
            final ServiceEvent service = mServices.get(position);

            holder.getBinding().text1.setText(service.getName());
            holder.getBinding().text2.setText(service.getType());

            holder.itemView.setOnClickListener(v -> {
                AppCompatActivity activity = (AppCompatActivity) v.getContext();
                activity.getSupportFragmentManager()
                    .beginTransaction()
    //                    .replace(R.id.main, ServiceListFragment.make(host))
                    .commit();
            });
        }

    }
}

File: flame/src/main/java/org/movieos/flame/fragments/ServiceListFragment.kt
package org.movieos.flame.fragments

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.movieos.flame.MyApplication
import org.movieos.flame.R
import org.movieos.flame.databinding.CellWithImageBinding
import org.movieos.flame.databinding.RecyclerViewFragmentBinding
import org.movieos.flame.models.FlameHost
import org.movieos.flame.utilities.BindingViewHolder
import org.movieos.flame.utilities.DiscoveryService
import timber.log.Timber
import java.util.*
import javax.jmdns.ServiceEvent

class ServiceListFragment : android.support.v4.app.Fragment() {

    private val adapter = ServiceListAdapter()
    var hostIdentifier: String? = null
    var binding: RecyclerViewFragmentBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        hostIdentifier = arguments.getString(HOST_IDENTIFIER)
        if (hostIdentifier == null) {
            Timber.e("no identifier in bundle")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val binding = RecyclerViewFragmentBinding.inflate(inflater, container, false)

        binding.toolbar.title = hostIdentifier
        val back = ContextCompat.getDrawable(context, R.drawable.ic_arrow_back_black_24dp)
        back.setTint(0xFFFFFFFF.toInt())
        binding.toolbar.navigationIcon = back
        binding.toolbar.setNavigationOnClickListener { activity.onBackPressed() }

        binding.recyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        binding.recyclerView.adapter = adapter

        val divider = DividerItemDecoration(activity, DividerItemDecoration.VERTICAL)
        divider.setDrawable(ContextCompat.getDrawable(context, R.drawable.divider))
        binding.recyclerView.addItemDecoration(divider)

        this.binding = binding
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (binding != null) {
            binding!!.unbind()
            binding = null
        }
    }

    override fun onResume() {
        super.onResume()
        onMessageEvent(null)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: DiscoveryService.HostsChangedNotification?) {
        val hosts = MyApplication.discoveryService.getHostsMatchingKey(hostIdentifier)
        if (!hosts.isEmpty()) {
            adapter.setServices(hosts[0].services)
        }
    }

    private class ServiceListAdapter internal constructor() : RecyclerView.Adapter<BindingViewHolder<CellWithImageBinding>>() {

        private val services = ArrayList<ServiceEvent>()

        internal fun setServices(services: List<ServiceEvent>) {
            this.services.clear()
            this.services.addAll(services)
            notifyDataSetChanged()
        }

        override fun getItemCount(): Int {
            return services.size
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindingViewHolder<CellWithImageBinding> {
            return BindingViewHolder.build(parent, R.layout.cell_with_image)
        }

        override fun onBindViewHolder(holder: BindingViewHolder<CellWithImageBinding>, position: Int) {
            val service = services[position]
            holder.binding.text1.text = service.name
            holder.binding.text2.text = service.type

            holder.itemView.setOnClickListener({ v ->
                val activity = v.context as AppCompatActivity
                activity.supportFragmentManager
                        .beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                        //                    .replace(R.id.main, ServiceListFragment.make(host))
                        .commit()
            })
        }

    }

    companion object {

        private val HOST_IDENTIFIER = "hostIdentifier"

        fun build(host: FlameHost): Fragment {
            val fragment = ServiceListFragment()
            fragment.arguments = Bundle()
            fragment.arguments.putString(HOST_IDENTIFIER, host.identifer)
            return fragment
        }
    }
}
